<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pedigree Analysis and Familial Aggregation • FamAgg</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Pedigree Analysis and Familial Aggregation">
<meta property="og:description" content="FamAgg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FamAgg</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.17.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/FamAgg.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/EuracBiomedicalResearch/FamAgg/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Pedigree Analysis and Familial Aggregation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/EuracBiomedicalResearch/FamAgg/blob/master/vignettes/FamAgg.Rmd"><code>vignettes/FamAgg.Rmd</code></a></small>
      <div class="hidden name"><code>FamAgg.Rmd</code></div>

    </div>

    
    
<pre><code>## Loading required package: kinship2</code></pre>
<pre><code>## Loading required package: Matrix</code></pre>
<pre><code>## Loading required package: quadprog</code></pre>
<pre><code>## Loading required package: igraph</code></pre>
<pre><code>## 
## Attaching package: 'igraph'</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     decompose, spectrum</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     union</code></pre>
<pre><code>## 
## Attaching package: 'FamAgg'</code></pre>
<pre><code>## The following object is masked from 'package:igraph':
## 
##     cliques</code></pre>
<pre><code>## The following object is masked from 'package:kinship2':
## 
##     pedigree</code></pre>
<p><strong>Package</strong>: <em><a href="https://bioconductor.org/packages/3.11/FamAgg">FamAgg</a></em><br><strong>Authors</strong>: J. Rainer, D. Taliun, C.X. Weichenberger<br><strong>Modified</strong>: 2020-05-06 16:50:58<br><strong>Compiled</strong>: Wed May 6 16:53:04 2020</p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This package provides basic pedigree analysis and plotting utilities as well as a variety of methods to evaluate familial clustering of cases from a given trait. Identification of families or groups of individuals within families with significant aggregation of cases can aid also in the selection of interesting and promising individuals for whole genome or exome sequencing projects.</p>
<p>For kinship coefficient calculations and pedigree plotting the package relies and extends the functionality of the <code>kinship2</code> package <span class="citation">[1]</span>.</p>
<p>If you use this package please cite Rainer et al. <span class="citation">[2]</span>.</p>
</div>
<div id="basic-pedigree-operations" class="section level1">
<h1 class="hasAnchor">
<a href="#basic-pedigree-operations" class="anchor"></a>Basic pedigree operations</h1>
<p>In the examples below we perform some simple pedigree operations, such as plotting the pedigree for an individual or family, finding the closest common ancestor for a set of individuals in a pedigree or retrieving the identifiers (IDs) of all ancestors for an individual. Basic pedigree information is stored in <code>FAData</code> objects, thus we first generate such an object from a subset of the Minnesota Breast Cancer Study provided by the <code>kinship2</code> package. In the example below, we generate the <code>FAData</code> providing a <code>data.frame</code> with the pedigree data, alternatively, the pedigree information could be imported from a file (see Section <a href="#org99a8d9e">3</a>). Upon data set creation the kinship matrix (i.e. a matrix containing the kinship coefficient between each pair of individuals in the whole pedigree) is internally calculated using the functionality from the <code>kinship2</code> package <span class="citation">[1]</span>.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">FamAgg</span>)

<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">minnbreast</span>)
<span class="co">## Subsetting to only few families of the whole data set.</span>
<span class="no">mbsub</span> <span class="kw">&lt;-</span> <span class="no">minnbreast</span>[<span class="no">minnbreast</span>$<span class="no">famid</span> <span class="kw">%in%</span> <span class="fl">4</span>:<span class="fl">14</span>, ]
<span class="no">mbped</span> <span class="kw">&lt;-</span> <span class="no">mbsub</span>[, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"famid"</span>, <span class="st">"id"</span>, <span class="st">"fatherid"</span>, <span class="st">"motherid"</span>, <span class="st">"sex"</span>)]
<span class="co">## Renaming column names.</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">mbped</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"family"</span>, <span class="st">"id"</span>, <span class="st">"father"</span>, <span class="st">"mother"</span>, <span class="st">"sex"</span>)
<span class="co">## Defining the optional argument age.</span>
<span class="no">endage</span> <span class="kw">&lt;-</span> <span class="no">mbsub</span>$<span class="no">endage</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">endage</span>) <span class="kw">&lt;-</span> <span class="no">mbsub</span>$<span class="no">id</span>
<span class="co">## Create the object.</span>
<span class="no">fad</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData.html">FAData</a></span>(<span class="kw">pedigree</span> <span class="kw">=</span> <span class="no">mbped</span>, <span class="kw">age</span> <span class="kw">=</span> <span class="no">endage</span>)</pre></body></html></div>
<p>We can access all the pedigree information stored in this object using the <code>pedigree</code> method, but also using <code><a href="../reference/FAData.html">$</a></code>. The row names of the pedigree <code>data.frame</code> as well as the names of the vectors returned by <code><a href="../reference/FAData.html">$</a></code> are the IDs of the individuals in the pedigree.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co">## Use the pedigree method to access the full pedigree</span>
<span class="co">## data.frame,</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAData.html">pedigree</a></span>(<span class="no">fad</span>))</pre></body></html></div>
<pre><code>##   family id father mother sex
## 1      4  1     NA     NA   M
## 2      4  2     NA     NA   F
## 3      4  3     25      4   F
## 4      4  4      1      2   F
## 5      4  5      1      2   M
## 6      4  6      1      2   M</code></pre>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co">## or access individual columns using $.</span>
<span class="co">## The ID of the father (0 representing "founders"):</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">fad</span>$<span class="no">father</span>)</pre></body></html></div>
<pre><code>##  1  2  3  4  5  6 
## NA NA 25  1  1  1</code></pre>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="co">## Mother:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">fad</span>$<span class="no">mother</span>)</pre></body></html></div>
<pre><code>##  1  2  3  4  5  6 
## NA NA  4  2  2  2</code></pre>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="co">## Sex:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">fad</span>$<span class="no">sex</span>)</pre></body></html></div>
<pre><code>## 1 2 3 4 5 6 
## M F F F M M 
## Levels: M F</code></pre>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="co">## We can also access the age of each individual, if</span>
<span class="co">## provided.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAData.html">age</a></span>(<span class="no">fad</span>))</pre></body></html></div>
<pre><code>##        1        2        3        4        5        6 
##       NA 78.05886 55.50000 48.00000 75.00342 53.63997</code></pre>
<p>To extract the pedigree for a single family we can use the <code>family</code> method, specifying either the ID of the family or the ID of an individual in the family.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="co">## Extract the pedigree information from family "4"...</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="fu"><a href="../reference/FAData.html">family</a></span>(<span class="no">fad</span>, <span class="kw">family</span> <span class="kw">=</span> <span class="fl">4</span>))</pre></body></html></div>
<pre><code>## [1] 43</code></pre>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAData.html">family</a></span>(<span class="no">fad</span>, <span class="kw">family</span> <span class="kw">=</span> <span class="fl">4</span>))</pre></body></html></div>
<pre><code>##   family id father mother sex
## 1      4  1     NA     NA   M
## 2      4  2     NA     NA   F
## 3      4  3     25      4   F
## 4      4  4      1      2   F
## 5      4  5      1      2   M
## 6      4  6      1      2   M</code></pre>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="co">## ...which is the same as extracting the family pedigree</span>
<span class="co">## for an individual of this family.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAData.html">family</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="fl">3</span>))</pre></body></html></div>
<pre><code>##   family id father mother sex
## 1      4  1     NA     NA   M
## 2      4  2     NA     NA   F
## 3      4  3     25      4   F
## 4      4  4      1      2   F
## 5      4  5      1      2   M
## 6      4  6      1      2   M</code></pre>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="co">## Note that IDs are internally always converted to character,</span>
<span class="co">## thus, using id=3 and id="3" return the same information.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAData.html">family</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="st">"3"</span>))</pre></body></html></div>
<pre><code>##   family id father mother sex
## 1      4  1     NA     NA   M
## 2      4  2     NA     NA   F
## 3      4  3     25      4   F
## 4      4  4      1      2   F
## 5      4  5      1      2   M
## 6      4  6      1      2   M</code></pre>
<p>Alternatively, we could subset the <code>FAData</code> to individuals of a single family.</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="co">## Subset the object to a single family.</span>
<span class="no">fam4</span> <span class="kw">&lt;-</span> <span class="no">fad</span>[<span class="no">fad</span>$<span class="no">family</span> <span class="kw">==</span> <span class="st">"4"</span>, ]
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">fam4</span>$<span class="no">family</span>)</pre></body></html></div>
<pre><code>## 
##  4 
## 43</code></pre>
<p>To explore this family we can plot its pedigree. By default, the plotting capabilities of the <code>kinship2</code> package are used to plot pedigrees, but alternatively, if all required dependencies are available, the <code>HaploPainter</code> <span class="citation">[3]</span> perl script (<a href="http://haplopainter.sourceforge.net/" class="uri">http://haplopainter.sourceforge.net/</a>) can be used instead. The <code>switchPlotfun</code> function can be used to switch the plotting back-end. Available arguments are <code>ks2paint</code> and <code>haplopaint</code> for <code>kinship2</code> and <code>HaploPainter</code> plotting, respectively. Note however, that <code>HaploPainter</code> only allows to export plots to a file, while <code>kinship2</code> plotting allows, in addition to export the plot, also to show it as a <em>standard</em> <code>R</code> plot.</p>
<p>Below we use the <code>switchPlotfun</code> to ensure the use of <code>kinship2</code> plotting (usually not required) and plot the full available pedigree of individual <code>3</code>. If the age of individuals is available, it will be plotted below the individual’s ID.</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="fu"><a href="../reference/plotting-functions.html">switchPlotfun</a></span>(<span class="st">"ks2paint"</span>)
<span class="co">## By supplying device="plot", we specify that we wish to visualize the</span>
<span class="co">## pedigree in an R plot. This is the default for "ks2paint", anyway.</span>
<span class="fu"><a href="../reference/FAData.html">plotPed</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="fl">3</span>, <span class="kw">device</span> <span class="kw">=</span> <span class="st">"plot"</span>)</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/plotPed-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The pedigree for an individual or a list of individuals can be extracted using the <code>buildPed</code> method. By default the method first tries to identify all parents up to 3 generations in the pedigree, and subsequently all children of the individuals and all identified parents.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="co">## Build the pedigree for individual 3.</span>
<span class="no">fullPed</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData.html">buildPed</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="st">"3"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">fullPed</span>)</pre></body></html></div>
<pre><code>## [1] 29</code></pre>
<p>Alternatively, we can extract the smallest possible pedigree for a list of individuals by specifying <code>prune=TRUE</code>. Internally, the function transforms the pedigree into a graph, tries to find all paths between the individuals and returns the sub-graph of all individuals along with individuals along the paths between them.</p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="co">## Find the subpedigree for individuals 21, 22 and 17.</span>
<span class="fu"><a href="../reference/FAData.html">buildPed</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">21</span>, <span class="fl">22</span>, <span class="fl">17</span>), <span class="kw">prune</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>##    family id father mother sex
## 3       4  3     25      4   F
## 4       4  4      1      2   F
## 1       4  1     NA     NA   M
## 8       4  8      1      2   F
## 17      4 17     28      8   M
## 21      4 21     24      3   M
## 22      4 22     24      3   F
## 2       4  2     NA     NA   F
## 25      4 25     NA     NA   M
## 28      4 28     NA     NA   M
## 24      4 24     NA     NA   M</code></pre>
<p>And the pedigree plot for that subset of the whole family:</p>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="fu"><a href="../reference/FAData.html">plotPed</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">21</span>, <span class="fl">22</span>, <span class="fl">17</span>), <span class="kw">prune</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/plotPed-3ids-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Note that the pedigree returned by the <code>buildPed</code> method for an individual might be different than the pedigree of a whole family. The pedigree returned by <code>buildPed</code> contains only individuals that share kinship with the specified individual. To exemplify this, we plot the pedigree for the family <code>14</code> in the Minnesota Breast Cancer data set. Note that the individuals in the pedigree plot depicted as diamonds are individuals with unknown gender. (The message “Did not plot…” is issued by the <code>kinship2</code> plotting function and indicates singletons that are assigned to the family but do neither have parents nor children.)</p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="fu"><a href="../reference/FAData.html">plotPed</a></span>(<span class="no">fad</span>, <span class="kw">family</span> <span class="kw">=</span> <span class="st">"14"</span>, <span class="kw">cex</span> <span class="kw">=</span> <span class="fl">0.4</span>)</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/plotPed-family-14-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code>## Did not plot the following people: 457 463 470 471 26067 26068 26098 26099</code></pre>
<p>In this family, founder <code>441</code> is the founder of two family branches. Building the pedigree for individual <code>440</code> will not include any of the individuals of the second branch, as he does not share kinship with any of them. The pedigree built for <code>447</code> on the other hand contains also individuals from the second branch as she shares kinship with them (<em>via</em> her mother <code>441</code>).</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="co">## Check if we have individual 26064 from the second branch in the pedigree</span>
<span class="co">## of individual 440.</span>
<span class="fu"><a href="https://rdrr.io/r/base/any.html">any</a></span>(<span class="fu"><a href="../reference/FAData.html">buildPed</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="st">"440"</span>)$<span class="no">id</span> <span class="kw">==</span> <span class="st">"26064"</span>)</pre></body></html></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="co">## What for the pedigree of 447?</span>
<span class="fu"><a href="https://rdrr.io/r/base/any.html">any</a></span>(<span class="fu"><a href="../reference/FAData.html">buildPed</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="st">"447"</span>)$<span class="no">id</span> <span class="kw">==</span> <span class="st">"26064"</span>)</pre></body></html></div>
<pre><code>## [1] TRUE</code></pre>
<p>A family pedigree may consist of many founder couples (i.e. individuals for which neither father nor mother is defined in the pedigree). To identify the pedigree’s founder couple (being the couple with the largest number of offspring generations in the pedigree) the <code>findFounders</code> method can be used. Note that the function returns always only one couple, even if there might be two founder couples in the family pedigree with the same number of offspring generations.</p>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="co">## Find founders for family 4.</span>
<span class="fu"><a href="../reference/Pedigree-utils.html">findFounders</a></span>(<span class="no">fad</span>, <span class="st">"4"</span>)</pre></body></html></div>
<pre><code>## [1] "1" "2"</code></pre>
<p>Alternatively, it might be of interest to determine the closest common ancestor between individuals in a pedigree. Below we use the <code>getCommonAncestor</code> method to identify the common ancestor for individuals <code>21</code>, <code>22</code> and <code>17</code> (which we know from the pedigree a bit above are <code>1</code> and <code>2</code>).</p>
<div class="sourceCode" id="cb46"><html><body><pre class="r"><span class="co">## Find the closest common ancestor.</span>
<span class="fu"><a href="../reference/Pedigree-utils.html">getCommonAncestor</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">21</span>, <span class="fl">22</span>, <span class="fl">17</span>))</pre></body></html></div>
<pre><code>## [1] "1" "2"</code></pre>
<p>Other useful methods are <code>getChildren</code>, <code>getAncestors</code> and <code>getSiblings</code>, that return the children (or all offspring generations up to a specified level), the parents (or all ancestors) or the siblings for the specified individuals, respectively.</p>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="co">## Get the children of ID 4.</span>
<span class="fu"><a href="../reference/Pedigree-utils.html">getChildren</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="st">"4"</span>, <span class="kw">max.generations</span> <span class="kw">=</span> <span class="fl">1</span>)</pre></body></html></div>
<pre><code>## [1] "3"</code></pre>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="co">## Get the offsprings.</span>
<span class="fu"><a href="../reference/Pedigree-utils.html">getChildren</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="st">"4"</span>)</pre></body></html></div>
<pre><code>## [1] "3"  "21" "22" "23"</code></pre>
<div class="sourceCode" id="cb52"><html><body><pre class="r"><span class="co">## Get all ancestors.</span>
<span class="fu"><a href="../reference/Pedigree-utils.html">getAncestors</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="st">"4"</span>)</pre></body></html></div>
<pre><code>## [1] "1" "2"</code></pre>
<div class="sourceCode" id="cb54"><html><body><pre class="r"><span class="co">## Get the siblings.</span>
<span class="fu"><a href="../reference/Pedigree-utils.html">getSiblings</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"4"</span>))</pre></body></html></div>
<pre><code>## [1] "4"  "5"  "6"  "7"  "8"  "9"  "10"</code></pre>
<p>In the whole Minnesota Breast Cancer data set there are 426 families corresponding to 426 founders that had cancer during the screening phase between 1944 and 1952. In the code block below we identify the affected founders per family.</p>
<div class="sourceCode" id="cb56"><html><body><pre class="r"><span class="co">## Add the trait information to the FAData object.</span>
<span class="no">cancer</span> <span class="kw">&lt;-</span> <span class="no">mbsub</span>$<span class="no">cancer</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">cancer</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">mbsub</span>$<span class="no">id</span>)
<span class="fu"><a href="../reference/FAData.html">trait</a></span>(<span class="no">fad</span>) <span class="kw">&lt;-</span> <span class="no">cancer</span>

<span class="co">## Identify the affected founders.</span>
<span class="co">## First all affected individuals.</span>
<span class="no">affIds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData.html">affectedIndividuals</a></span>(<span class="no">fad</span>)
<span class="co">## Identify founders for each family.</span>
<span class="no">founders</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">fad</span>$<span class="no">family</span>), <span class="kw">function</span>(<span class="no">z</span>){
    <span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="fu"><a href="../reference/Pedigree-utils.html">findFounders</a></span>(<span class="no">fad</span>, <span class="kw">family</span> <span class="kw">=</span> <span class="no">z</span>))
})
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">founders</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">fad</span>$<span class="no">family</span>)

<span class="co">## Track the affected founder.</span>
<span class="no">affFounders</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">founders</span>, <span class="kw">function</span>(<span class="no">z</span>){
    <span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="no">z</span>[<span class="no">z</span> <span class="kw">%in%</span> <span class="no">affIds</span>])
})
<span class="co">## Interestingly, not all founders are affected! It seems in some cases</span>
<span class="co">## parents of the affected participants in the screening phase have also</span>
<span class="co">## been included.</span>
<span class="no">affFounders</span> <span class="kw">&lt;-</span> <span class="no">affFounders</span>[<span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">affFounders</span>, <span class="no">length</span>)) <span class="kw">&gt;</span> <span class="fl">0</span>]

<span class="co">## The number of families analyzed.</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">founders</span>)</pre></body></html></div>
<pre><code>## [1] 10</code></pre>
<div class="sourceCode" id="cb58"><html><body><pre class="r"><span class="co">## The number of families with affected founder.</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">affFounders</span>)</pre></body></html></div>
<pre><code>## [1] 2</code></pre>
<p>Unexpectedly, only in few families one of the founders is affected. For the other families additional (unaffected) ancestors might have been added at a later time point.</p>
<p>Next we get the number of affected individuals that are related to these affected founders.</p>
<div class="sourceCode" id="cb60"><html><body><pre class="r"><span class="no">kin2affFounders</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-utils.html">shareKinship</a></span>(<span class="no">fad</span>, <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">affFounders</span>))

<span class="co">## How many of these are affected?</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">kin2affFounders</span> <span class="kw">%in%</span> <span class="no">affIds</span>)</pre></body></html></div>
<pre><code>## [1] 7</code></pre>
<div class="sourceCode" id="cb62"><html><body><pre class="r"><span class="co">## How many affected are not related to an affected founder?</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(!(<span class="no">affIds</span> <span class="kw">%in%</span> <span class="no">kin2affFounders</span>))</pre></body></html></div>
<pre><code>## [1] 21</code></pre>
<div id="pedigree-analysis-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#pedigree-analysis-methods" class="anchor"></a>Pedigree analysis methods</h2>
<p>In this section we perform some more advanced pedigree operations. First, we identify all individuals in the pedigree that share kinship with individual <code>4</code>.</p>
<div class="sourceCode" id="cb64"><html><body><pre class="r"><span class="co">## Get all individuals sharing kinship with individual 4.</span>
<span class="fu"><a href="../reference/Pedigree-utils.html">shareKinship</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="st">"4"</span>)</pre></body></html></div>
<pre><code>##  [1] "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14" "15"
## [16] "16" "17" "18" "19" "20" "21" "22" "23"</code></pre>
<p>Next, we determine generations within the pedigree. Generations can only be estimated for a single family, since in most instances e.g. the year of birth is not available. Thus, generations are estimated considering the relation between individuals, starting from the founder couple, i.e. generation 0, assigning generation 1 to their children and all the mates of their children and so on. The <code>estimateGenerations</code> method calculates such generation numbers for each family defined in the object (or for a single family, if the family ID is provided). The result is returned as a list with the list names corresponding to the family ID and the list elements being the estimated generation numbers (with names corresponding to the ID of the respective individual).</p>
<div class="sourceCode" id="cb66"><html><body><pre class="r"><span class="co">## Estimate generation levels for all families.</span>
<span class="fu"><a href="../reference/Pedigree-utils.html">estimateGenerations</a></span>(<span class="no">fad</span>)[<span class="fl">1</span>:<span class="fl">3</span>]</pre></body></html></div>
<pre><code>## $`4`
##  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 
##  0  0  2  1  1  1  1  1  1  1  2  2  2  2  2  2  2  2  2  2  3  3  3  2  1  1 
## 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 
##  1  1  1 NA NA NA NA NA NA NA NA NA NA NA NA NA NA 
## 
## $`5`
##    44    45    46    47    48    49    50    51    52    53    54    55    56 
##     0     0     2     2     2     2     2     2     2     2     1     3     3 
##    57    58    59    60    61    62    63    64    65    66    67    68    69 
##     3     3     3     3     3     3     3     3     2     2     2    NA     2 
##    70    71    72    73    74    75    76    77    78    79 26050 26051 
##     1    NA    NA    NA    NA    NA    NA    NA    NA     2    NA    NA 
## 
## $`6`
##    80    81    82    83    84    85    86    87    88    89    90    91    92 
##     0     0     2     2     1     1     1     1     1     1     1     1     1 
##    93    94    95    96    97    98    99   100   101   102   103   104   105 
##     1     2     2     2     2     2     2     2     3     3     3     3     2 
##   106   107   108   109   110   111   112   113   114   115   116   117   118 
##     2     1     1     1    NA    NA    NA    NA     2    NA    NA    NA    NA 
## 26052 26053 
##     3     3</code></pre>
<p>Individuals without generation level (i.e. with an <code>NA</code>) are not connected to any other individual in the pedigree (and thus most likely represent errors in the pedigree).</p>
<p>In addition, it is also possible to calculate generation levels relative to a (single) specified individual:</p>
<div class="sourceCode" id="cb68"><html><body><pre class="r"><span class="no">gens</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-utils.html">generationsFrom</a></span>(<span class="no">fad</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="st">"4"</span>)</pre></body></html></div>
<p>We can render these generation numbers into the pedigree:</p>
<div class="sourceCode" id="cb69"><html><body><pre class="r"><span class="fu"><a href="../reference/FAData.html">plotPed</a></span>(<span class="no">fad</span>, <span class="kw">family</span> <span class="kw">=</span> <span class="fl">4</span>, <span class="kw">label2</span> <span class="kw">=</span> <span class="no">gens</span>)</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/plotPed-with-generations-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code>## Did not plot the following people: 30 31 32 33 34 35 36 37 38 39 40 41 42 43</code></pre>
</div>
<div id="additional-plotting-options" class="section level2">
<h2 class="hasAnchor">
<a href="#additional-plotting-options" class="anchor"></a>Additional plotting options</h2>
<p>If a trait information is available it might be of interest to highlight affected individuals in the pedigree. Trait information should always be coded as <code>0</code> (or <code>FALSE</code>) for unaffected and <code>1</code> (or <code>TRUE</code>) for affected. In the example below, we use the <em>cancer</em> information from the Minnesota Breast Cancer Study.</p>
<div class="sourceCode" id="cb71"><html><body><pre class="r"><span class="co">## Extract the cancer trait information.</span>
<span class="no">tcancer</span> <span class="kw">&lt;-</span> <span class="no">mbsub</span>$<span class="no">cancer</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">tcancer</span>) <span class="kw">&lt;-</span> <span class="no">mbsub</span>$<span class="no">id</span>
<span class="co">## Set the trait.</span>
<span class="fu"><a href="../reference/FAData.html">trait</a></span>(<span class="no">fad</span>) <span class="kw">&lt;-</span> <span class="no">tcancer</span></pre></body></html></div>
<p>We can now extract the trait information from the object or identify directly the phenotyped or affected individuals.</p>
<div class="sourceCode" id="cb72"><html><body><pre class="r"><span class="co">## Extract the trait information.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAData.html">trait</a></span>(<span class="no">fad</span>))</pre></body></html></div>
<pre><code>## 1 2 3 4 5 6 
## 0 0 0 1 0 0</code></pre>
<div class="sourceCode" id="cb74"><html><body><pre class="r"><span class="co">## We can also extract the IDs of the affected individuals.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAData.html">affectedIndividuals</a></span>(<span class="no">fad</span>))</pre></body></html></div>
<pre><code>## [1] "4"   "11"  "37"  "54"  "84"  "122"</code></pre>
<div class="sourceCode" id="cb76"><html><body><pre class="r"><span class="co">## Or the IDs of the phenotyped individuals.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAData.html">phenotypedIndividuals</a></span>(<span class="no">fad</span>))</pre></body></html></div>
<pre><code>## [1] "1" "2" "3" "4" "5" "6"</code></pre>
<p>Plotting a <code>FAData</code> object with trait information results in a pedigree plot with highlighted affected individuals (for <code>kinship2</code> pedigree plotting: affected, unaffected and not phenotyped are represented as filled symbols, open symbols and symbols with a question mark inside, respectively).</p>
<div class="sourceCode" id="cb78"><html><body><pre class="r"><span class="co">## Plotting the pedigree for family "9".</span>
<span class="fu"><a href="../reference/FAData.html">plotPed</a></span>(<span class="no">fad</span>, <span class="kw">family</span> <span class="kw">=</span> <span class="st">"9"</span>)</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/plotPed-with-trait-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code>## Did not plot the following people: 200 204 206 210 212 214 215 216 217 219</code></pre>
<p>In addition, we can manually highlight individuals using the <code>highlight.ids</code> argument. For <code>kinship2</code> pedigree plotting, a list of length 2 is supported as argument <code>highlight.ids</code>, with the first element being plotted on the top left corner of the symbol and the second element on the top right corner.</p>
<div class="sourceCode" id="cb80"><html><body><pre class="r"><span class="co">## Plotting the pedigree for family "9".</span>
<span class="fu"><a href="../reference/FAData.html">plotPed</a></span>(<span class="no">fad</span>, <span class="kw">family</span> <span class="kw">=</span> <span class="st">"9"</span>, <span class="kw">highlight.ids</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">a</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"185"</span>, <span class="st">"201"</span>, <span class="st">"198"</span>),
                        <span class="kw">b</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"193"</span>)))</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/plotPed-trait-highlight-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code>## Did not plot the following people: 200 204 206 210 212 214 215 216 217 219</code></pre>
<p>An alternative way to highlight individuals or add text to the plot is to use the arguments <code>label1</code>, <code>label2</code> and <code>label3</code> or the <code>plotPed</code> method.</p>
</div>
<div id="graph-utilities" class="section level2">
<h2 class="hasAnchor">
<a href="#graph-utilities" class="anchor"></a>Graph utilities</h2>
<p>Pedigrees can also be transformed to graphs using the <code>ped2graph</code> function. That way all graph theory methods implemented in e.g. the <code>igraph</code> package can be applied to pedigrees.</p>
<div class="sourceCode" id="cb82"><html><body><pre class="r"><span class="co">## Transform the full pedigree to a graph.</span>
<span class="no">fullGraph</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-utils.html">ped2graph</a></span>(<span class="fu"><a href="../reference/FAData.html">pedigree</a></span>(<span class="no">fad</span>))

<span class="co">## In addition, build the graph for a single family.</span>
<span class="no">singleFam</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-utils.html">ped2graph</a></span>(<span class="fu"><a href="../reference/FAData.html">family</a></span>(<span class="no">fad</span>, <span class="kw">family</span><span class="kw">=</span><span class="fl">4</span>))</pre></body></html></div>
<p>We can plot these pedigrees also as graph and could use any of the layout methods provided in the <code>igraph</code> package.</p>
<div class="sourceCode" id="cb83"><html><body><pre class="r"><span class="co">## Build the layout.</span>
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">fullGraph</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="FamAgg_files/figure-html/plot-igraph-1.png" alt="Pedigrees represented as graphs." width="700"><p class="caption">
Pedigrees represented as graphs.
</p>
</div>
<div class="sourceCode" id="cb84"><html><body><pre class="r"><span class="no">lay</span> <span class="kw">&lt;-</span> <span class="fu">layout_</span>(<span class="no">singleFam</span>, <span class="fu">on_grid</span>())
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">singleFam</span>, <span class="kw">layout</span> <span class="kw">=</span> <span class="no">lay</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="FamAgg_files/figure-html/plot-igraph-2.png" alt="Pedigrees represented as graphs." width="700"><p class="caption">
Pedigrees represented as graphs.
</p>
</div>
<p>The <code>connectedSubgraph</code> function implemented in the <code>FamAgg</code> package provides additional functionality to find the smallest connected subgraph of a list of submitted nodes (i.e. individuals).</p>
<p>In the code below we want to extract the smallest possible connected subgraph of the pedigree-graph of family 4 containing individuals <code>7</code>, <code>8</code>, <code>27</code> and <code>17</code>.</p>
<div class="sourceCode" id="cb85"><html><body><pre class="r"><span class="no">subgr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-utils.html">connectedSubgraph</a></span>(<span class="no">singleFam</span>, <span class="kw">nodes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"7"</span>, <span class="st">"8"</span>, <span class="st">"27"</span>, <span class="st">"17"</span>))</pre></body></html></div>
<p>This is in principle what the <code>buildPed</code> method with the option <code>prune=TRUE</code> does to find the smallest pedigree for a set of individuals, only that <code>buildPed</code> ensures that also eventually missing parents are added.</p>
<div class="sourceCode" id="cb86"><html><body><pre class="r"><span class="co">## Plot the graph.</span>
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">subgr</span>)</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/plot-subgraph,%20subgraph-plot-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb87"><html><body><pre class="r"><span class="co">## Similar to buildPed/plotPed with prune=TRUE.</span>
<span class="fu"><a href="../reference/FAData.html">plotPed</a></span>(<span class="no">fad</span>, <span class="kw">id</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"7"</span>, <span class="st">"8"</span>, <span class="st">"17"</span>, <span class="st">"27"</span>), <span class="kw">prune</span><span class="kw">=</span><span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>## Removing singletons... none present.</code></pre>
<p><img src="FamAgg_files/figure-html/plot-subgraph,%20subgraph-plot-2.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="importing-and-exporting-pedigree-data" class="section level1">
<h1 class="hasAnchor">
<a href="#importing-and-exporting-pedigree-data" class="anchor"></a>Importing and exporting pedigree data<a id="org99a8d9e"></a>
</h1>
<p>Besides providing the pedigree data as a <code>data.frame</code>, the <code>FAData</code> constructor can also read pedigree data from various file formats, such as plink <span class="citation">[4]</span> <em>ped</em> or <em>fam</em> files (<a href="http://pngu.mgh.harvard.edu/~purcell/plink/data.shtml" class="uri">http://pngu.mgh.harvard.edu/~purcell/plink/data.shtml</a>) or generic text files.</p>
<div class="sourceCode" id="cb89"><html><body><pre class="r"><span class="co">## Import a "ped" file.</span>
<span class="no">pedFile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"txt/minnbreastsub.ped.gz"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"FamAgg"</span>)
<span class="co">## Quick glance at the file.</span>
<span class="fu"><a href="https://rdrr.io/r/base/readLines.html">readLines</a></span>(<span class="no">pedFile</span>, <span class="kw">n</span> <span class="kw">=</span> <span class="fl">1</span>)</pre></body></html></div>
<pre><code>## [1] "4\t1\t0\t0\t1\t1"</code></pre>
<div class="sourceCode" id="cb91"><html><body><pre class="r"><span class="no">fad</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData.html">FAData</a></span>(<span class="no">pedFile</span>)

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAData.html">pedigree</a></span>(<span class="no">fad</span>))</pre></body></html></div>
<pre><code>##   family id father mother sex
## 1      4  1   &lt;NA&gt;   &lt;NA&gt;   M
## 2      4  2   &lt;NA&gt;   &lt;NA&gt;   F
## 3      4  3     25      4   F
## 4      4  4      1      2   F
## 5      4  5      1      2   M
## 6      4  6      1      2   M</code></pre>
<p>Alternatively, we can import pedigree data from generic input files.</p>
<div class="sourceCode" id="cb93"><html><body><pre class="r"><span class="co">## Create the FAData by reading data from a txt file.</span>
<span class="no">pedFile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"txt/minnbreastsub.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"FamAgg"</span>)
<span class="no">fad</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData.html">FAData</a></span>(<span class="kw">pedigree</span> <span class="kw">=</span> <span class="no">pedFile</span>, <span class="kw">header</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">id.col</span> <span class="kw">=</span> <span class="st">"id"</span>,
          <span class="kw">family.col</span> <span class="kw">=</span> <span class="st">"famid"</span>, <span class="kw">father.col</span> <span class="kw">=</span> <span class="st">"fatherid"</span>,
          <span class="kw">mother.col</span> <span class="kw">=</span> <span class="st">"motherid"</span>)</pre></body></html></div>
<p>And we can export pedigree data again using the <code>export</code> method. In the example below, we subset the whole pedigree to the pedigree of family 4 and export this as a <em>ped</em> file.</p>
<div class="sourceCode" id="cb94"><html><body><pre class="r"><span class="no">tmpF</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span>()

<span class="co">## Subset the pedigree to family 4</span>
<span class="no">fam4</span> <span class="kw">&lt;-</span> <span class="no">fad</span>[<span class="no">fad</span>$<span class="no">family</span> <span class="kw">==</span> <span class="fl">4</span>, ]

<span class="co">## Export data in ped format.</span>
<span class="fu"><a href="../reference/FAData.html">export</a></span>(<span class="no">fam4</span>, <span class="no">tmpF</span>, <span class="kw">format</span> <span class="kw">=</span> <span class="st">"ped"</span>)</pre></body></html></div>
</div>
<div id="testing-for-familial-aggregation" class="section level1">
<h1 class="hasAnchor">
<a href="#testing-for-familial-aggregation" class="anchor"></a>Testing for familial aggregation</h1>
<p>Familial aggregation aims to identify families within large ancestral pedigrees that show a non-random aggregation of traits.</p>
<p>As an example, we analyze here data from the Minnesota Breast Cancer Record, which is provided by the <code>kinship2</code> package. In brief, this data set consists of genealogical information from 426 unrelated founders diagnosed with breast cancer whose families entered a longitudinal study on cancer in the state of Minnesota (USA) in 1944. Cancer cases are encoded with a <code>1</code> in column <code>cancer</code> in the <code>minnbreast</code> <code>data.frame</code>. Note however that, besides breast cancer, also prostate cancer cases are reported. This unfortunately causes a systematic bias in the data set as families were only included if a founder was diagnosed with breast cancer, but all occurrences of both breast and prostate cancer are reported. Based on this bias many of the results below should be taken with caution. Another important information is provided in column <code>endage</code>, which represents either the age of cancer onset, the age at the end of the study or the age at death of the participant.</p>
<p>Note that, to reduce computation time, we perform the analysis only on a subset of families from the Minnesota Breast Cancer record and reduce the number of simulation runs. We specifically selected some families with a high percentage of cancer cases, thus, the analysis presented here is biased. Also, in a real analysis you should increase the <code>nsim</code> argument.</p>
<div class="sourceCode" id="cb95"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">FamAgg</span>)
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">18011977</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">minnbreast</span>)
<span class="co">## Subset the dataset to reduce processing time.</span>
<span class="no">mbsub</span> <span class="kw">&lt;-</span> <span class="no">minnbreast</span>[<span class="no">minnbreast</span>$<span class="no">famid</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">4</span>:<span class="fl">100</span>, <span class="fl">173</span>, <span class="fl">432</span>), ]
<span class="co">## Uncomment the line below to use the whole dataset instead.</span>
<span class="co">## mbsub &lt;- minnbreast</span>

<span class="co">## Define the number of simulations we perform.</span>
<span class="co">## nsim &lt;- 10000</span>
<span class="no">nsim</span> <span class="kw">&lt;-</span> <span class="fl">1000</span>

<span class="no">mbped</span> <span class="kw">&lt;-</span> <span class="no">mbsub</span>[, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"famid"</span>, <span class="st">"id"</span>, <span class="st">"fatherid"</span>, <span class="st">"motherid"</span>, <span class="st">"sex"</span>)]
<span class="co">## Renaming column names.</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">mbped</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"family"</span>, <span class="st">"id"</span>, <span class="st">"father"</span>, <span class="st">"mother"</span>, <span class="st">"sex"</span>)
<span class="co">## Create the FAData object.</span>
<span class="no">fad</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData.html">FAData</a></span>(<span class="kw">pedigree</span> <span class="kw">=</span> <span class="no">mbped</span>)

<span class="co">## Define the trait.</span>
<span class="no">tcancer</span> <span class="kw">&lt;-</span> <span class="no">mbsub</span>$<span class="no">cancer</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">tcancer</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">mbsub</span>$<span class="no">id</span>)</pre></body></html></div>
<p>In the following section we analyze the data set first using the <em>genealogical index</em> method <span class="citation">[5]</span> (Section <a href="#org1dbf9f5">4.1</a>), then we estimate the per-individual risk of disease using the <em>familial incidence rate</em> (FIR, also abbreviated as <em>FR</em> in the original work) <span class="citation">[6]</span> (Section <a href="#org7913823">4.2</a>) and apply our <em>kinship sum test</em> to identify affected individuals exhibiting a higher relationship to other affected individuals than what would be expected by chance (Section <a href="#orgb323644">4.3</a>). Subsequently, we apply our <em>kinship group test</em> (Section <a href="#orgaac6b02">4.4</a>) that allows to identify highly clustered affected individuals within families.</p>
<p>The <em>genealogical index of familiality</em> and the <em>familial incidence rate</em> test are well established methods while the <em>kinship sum test</em> and the <em>kinship group test</em> are novel approaches presented here for the first time.</p>
<div id="genealogical-index-of-familiality" class="section level2">
<h2 class="hasAnchor">
<a href="#genealogical-index-of-familiality" class="anchor"></a><em>Genealogical index of familiality</em> <a id="org1dbf9f5"></a>
</h2>
<p>We next calculate the <em>genealogical index of familiality</em> (GIF) <span class="citation">[5]</span> (referred to as the <em>genealogical index</em> in the original work) for cancer occurrence in a subset of the Minnesota Breast Cancer Record data set. For a given trait (e.g. whether or not an individual was diagnosed with a certain type of cancer), the method computes the mean kinship between affected individuals (cases) in the whole pedigree along with mean kinship values of randomly drawn sets of individuals. The distribution of average kinship values among the control sets is then used to estimate the probability that the observed level of kinship among the cases is due to chance.</p>
<p>Below, we perform the analysis using the <code>genealogicalIndexTest</code> method on the <code>cancer</code> trait. In its default setting, the <code>genealogicalIndexTest</code> function uses all phenotyped individuals in the pedigree as control population from which sets of random samples equal in size to the number of affected are drawn.</p>
<p>Note that by default the function excludes all singletons (i.e. unconnected individuals in the pedigree) from the analysis. Changing the argument <code>rm.singletons</code> to <code>FALSE</code> will estimate the GIF on the full data set.</p>
<div class="sourceCode" id="cb96"><html><body><pre class="r"><span class="co">## Calculate the genealogical index of familiality.</span>
<span class="no">gi</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData-analysis.html">genealogicalIndexTest</a></span>(<span class="no">fad</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="no">tcancer</span>,
                <span class="kw">traitName</span> <span class="kw">=</span> <span class="st">"cancer"</span>, <span class="kw">nsim</span> <span class="kw">=</span> <span class="no">nsim</span>)

<span class="co">## Display the result.</span>
<span class="fu"><a href="../reference/FAProbResults.html">result</a></span>(<span class="no">gi</span>)</pre></body></html></div>
<pre><code>##   trait_name total_phenotyped total_affected entity_id entity_ctrls
## 1     cancer             3508            248         1         2804
##   entity_affected genealogical_index pvalue  padj
## 1             214           192.5102  0.001 0.001</code></pre>
<p>The column <em>genealogical index</em> of the result <code>data.frame</code> shown above represents the mean kinship between all pairs of affected individuals in the pedigree multiplied by <code>100000</code> for easier interpretation. Thus, according to the GIF test, a clustering of cancer cases is present in the analyzed pedigree. The output messages from the method call indicate that some individuals have been excluded from the test since they were either not phenotyped in the trait (i.e. have a missing value in trait), or are not <em>connected</em> in the family pedigree (do not share kinship with any other individual in the pedigree after removing non-phenotyped individuals).</p>
<p>The genealogical index of familiality implementation in this package adds some more flexibility to the original approach. The definition of the appropriate set of control individuals from which random samples are drawn can be specified with the <code>controlSetMethod</code> argument. Also, it is possible to perform a stratified sampling, e.g. if the group of affected cases in a pedigree consists of 5 female and 3 male individuals, submitting the sex of each individual in the pedigree with the argument <code>strata</code> (i.e. <code>strata=fad$sex</code>, with <code>fad</code> being the <code>FAData</code> object on which the analysis is performed) allows the function to define random control sets with the same proportion of male/female individuals.</p>
<p>In the next example, we use the <code>getSexMatched</code> function to define the set of control individuals and also the <code>getExternalMatched</code> submitting the gender information of each individual. The results from both approaches are essentially identical, and in the present data set not that useful, as the Minnesota Breast Cancer data set lists both, breast cancer and prostate cancer in column <code>cancer</code>, thus, the set of control individuals will contain all individuals with known sex.</p>
<div class="sourceCode" id="cb98"><html><body><pre class="r"><span class="co">## Calculate the genealogical index of familiality using random sampling from</span>
<span class="co">## a sex matched control set.</span>
<span class="no">giSexMatch</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData-analysis.html">genealogicalIndexTest</a></span>(<span class="no">fad</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="no">tcancer</span>,
                    <span class="kw">traitName</span> <span class="kw">=</span> <span class="st">"cancer"</span>, <span class="kw">nsim</span> <span class="kw">=</span> <span class="no">nsim</span>,
                    <span class="kw">controlSetMethod</span> <span class="kw">=</span> <span class="st">"getSexMatched"</span>)

<span class="co">## Use an external vector to perform the matching.</span>
<span class="co">## The results are essentially identical.</span>
<span class="no">giExtMatch</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData-analysis.html">genealogicalIndexTest</a></span>(<span class="no">fad</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="no">tcancer</span>,
                    <span class="kw">traitName</span> <span class="kw">=</span> <span class="st">"cancer"</span>, <span class="kw">nsim</span> <span class="kw">=</span> <span class="no">nsim</span>,
                    <span class="kw">controlSetMethod</span> <span class="kw">=</span> <span class="st">"getExternalMatched"</span>,
                    <span class="kw">match.using</span> <span class="kw">=</span> <span class="no">fad</span>$<span class="no">sex</span>)</pre></body></html></div>
<p>Note that any matching or stratified sampling can lead to the exclusion of individuals with missing values in either the matching criteria or the strata.</p>
<p>In the Minnesota Breast Cancer data set, the number of prostate cancer cases is much lower than the number of breast cancer cases, thus, simple random sampling might result in an biased genealogical index of familiality estimate since about the same proportion of male and female individuals will be sampled. To account for such cases a stratified sampling, as performed below, can be used instead.</p>
<div class="sourceCode" id="cb99"><html><body><pre class="r"><span class="co">## Evaluate the proportion of male and femal cases.</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">gi</span>$<span class="no">sex</span>[<span class="fu"><a href="../reference/FAData.html">affectedIndividuals</a></span>(<span class="no">gi</span>)])</pre></body></html></div>
<pre><code>## 
##   M   F 
##  39 206</code></pre>
<div class="sourceCode" id="cb101"><html><body><pre class="r"><span class="co">## We can use the gender information to perform stratified sampling, i.e.</span>
<span class="co">## in each permutation a random set of 3 male and 15 females will be selected.</span>
<span class="no">giStrata</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData-analysis.html">genealogicalIndexTest</a></span>(<span class="no">fad</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="no">tcancer</span>,
                  <span class="kw">traitName</span> <span class="kw">=</span> <span class="st">"cancer"</span>, <span class="kw">nsim</span> <span class="kw">=</span> <span class="no">nsim</span>,
                  <span class="kw">strata</span> <span class="kw">=</span> <span class="no">fad</span>$<span class="no">sex</span>)

<span class="fu"><a href="../reference/FAProbResults.html">result</a></span>(<span class="no">giStrata</span>)</pre></body></html></div>
<pre><code>##   trait_name total_phenotyped total_affected entity_id entity_ctrls
## 1     cancer             3508            248         1         2801
##   entity_affected genealogical_index pvalue  padj
## 1             214           192.5102  0.001 0.001</code></pre>
<p>Finally, we plot the result from the simulation. The blue vertical line in the plot below represents the mean kinship value between all affected individuals in the pedigree. The distribution of mean kinship values from the 1000 randomly drawn sets are shown in grey color.</p>
<div class="sourceCode" id="cb103"><html><body><pre class="r"><span class="co">## Plot the result.</span>
<span class="fu"><a href="../reference/FAGenIndexResults.html">plotRes</a></span>(<span class="no">giStrata</span>)</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/gif-4-plot,%20mbreast-genealogical-index-result-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The genealogical index of familiality can also be estimated by the <code>gif</code> function from the <code>gap</code> R-package. Below we calculate the estimate using both methods and compare the resulting estimate. Note that the <code>gif</code> method reports only the genealogical index of familiality estimate but does not estimate significance.</p>
<div class="sourceCode" id="cb104"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">gap</span>)

<span class="co">## Adding the trait information, so the extracted pedigree data.frame will</span>
<span class="co">## also contain a column "affected" with that information.</span>
<span class="fu"><a href="../reference/FAData.html">trait</a></span>(<span class="no">fad</span>) <span class="kw">&lt;-</span> <span class="no">tcancer</span>

<span class="co">## Extract the pedigree and re-format it for the gif function.</span>
<span class="no">pedi</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData.html">pedigree</a></span>(<span class="no">fad</span>)
<span class="co">## Remove singletons.</span>
<span class="no">pedi</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-utils.html">removeSingletons</a></span>(<span class="no">pedi</span>)
<span class="no">pedi</span>[<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">pedi</span>$<span class="no">father</span>), <span class="st">"father"</span>] <span class="kw">&lt;-</span> <span class="fl">0</span>
<span class="no">pedi</span>[<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">pedi</span>$<span class="no">mother</span>), <span class="st">"mother"</span>] <span class="kw">&lt;-</span> <span class="fl">0</span>

<span class="co">## Identify the affected individuals.</span>
<span class="no">affIds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="no">pedi</span>$<span class="no">id</span>[<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">pedi</span>$<span class="no">affected</span> <span class="kw">==</span> <span class="fl">1</span>)])

<span class="co">## Execute the gif method contained in the gap package.</span>
<span class="no">gifRes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gap/man/gif.html">gif</a></span>(<span class="no">pedi</span>[, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"id"</span>, <span class="st">"father"</span>, <span class="st">"mother"</span>)], <span class="no">affIds</span>)

<span class="co">## Calculate the GIF using FamAgg's genealogicalIndexTest.</span>
<span class="no">gifT</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData-analysis.html">genealogicalIndexTest</a></span>(<span class="no">fad</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="no">tcancer</span>, <span class="kw">nsim</span> <span class="kw">=</span> <span class="fl">100</span>)

<span class="co">## Comparing the results:</span>
<span class="no">gifRes</span><span class="kw">[[</span><span class="fl">1</span>]] <span class="kw">==</span> <span class="fu"><a href="../reference/FAProbResults.html">result</a></span>(<span class="no">gifT</span>)$<span class="no">genealogical_index</span></pre></body></html></div>
<pre><code>## [1] FALSE</code></pre>
<p>Thus, the GIF estimate from the <code>gap</code> package is identical to the one from the <code>FamAgg</code> package.</p>
<p>In the examples above, we tested for an enrichment of cancer cases in the full data set, i.e. across all families. In addition, we can perform the test individually for each family, by setting the <code>perFamilyTest</code> parameter of the <code>genealogicalIndexTest</code> to <code>TRUE</code>, and thus test for a clustering of cancer cases within each family.</p>
<div class="sourceCode" id="cb106"><html><body><pre class="r"><span class="co">## Perform the analysis (no strata etc) separately for each family.</span>
<span class="no">giFam</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData-analysis.html">genealogicalIndexTest</a></span>(<span class="no">fad</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="no">tcancer</span>, <span class="kw">nsim</span> <span class="kw">=</span> <span class="no">nsim</span>,
                   <span class="kw">perFamilyTest</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                   <span class="kw">traitName</span> <span class="kw">=</span> <span class="st">"Cancer"</span>)

<span class="co">## Display the result from the analysis.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAProbResults.html">result</a></span>(<span class="no">giFam</span>))</pre></body></html></div>
<pre><code>##     trait_name total_phenotyped total_affected entity_id entity_ctrls
## 13      Cancer             3508            248        13           29
## 432     Cancer             3508            248       432          106
## 14      Cancer             3508            248        14           31
## 89      Cancer             3508            248        89           78
## 30      Cancer             3508            248        30           25
## 40      Cancer             3508            248        40           39
##     entity_affected genealogical_index pvalue   padj
## 13                5          21250.000  0.001 0.0510
## 432              15           9940.476  0.002 0.0510
## 14                5          21250.000  0.003 0.0510
## 89                5          15625.000  0.028 0.3315
## 30                3          25000.000  0.037 0.3315
## 40                3          20833.333  0.039 0.3315</code></pre>
</div>
<div id="familial-incidence-rate-fir" class="section level2">
<h2 class="hasAnchor">
<a href="#familial-incidence-rate-fir" class="anchor"></a><em>Familial incidence rate</em> (FIR)<a id="org7913823"></a>
</h2>
<p>A per-individual risk of e.g. disease can be calculated using the <em>familial incidence rate</em> (FIR, abbreviated as <em>FR</em> in the original work) <span class="citation">[6]</span>. This measure considers the kinship of each individual with any affected in a given trait in the pedigree and the time at risk for each individual. Thus, the FIR is an estimate for the risk per gene-time for each individual given the disease-experience in the cohort.</p>
<p>As <em>time at risk</em> for each individual we use the <code>endage</code> column in the Minnesota Breast Cancer data set, which represents the participant’s age at the last follow-up or at cancer incidence. This estimate of time at risk is rather crude and in a real life situation a better, more accurate, estimate that is based e.g. on the birth dates and dates of last follow up or incidence might be used instead. See the help of functions <code>estimateTimeAtRisk</code> and <code>sliceAge</code> for details and options related to <em>time at risk</em>.</p>
<div class="sourceCode" id="cb108"><html><body><pre class="r"><span class="co">## Estimate the risk for each individual using the familial incidence</span>
<span class="co">## rate method. We use the "endage" provided in the Minnesota Breast Cancer</span>
<span class="co">## Record as a measure for time at risk.</span>
<span class="no">fr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData-analysis.html">familialIncidenceRate</a></span>(<span class="no">fad</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="no">tcancer</span>, <span class="kw">timeAtRisk</span> <span class="kw">=</span> <span class="no">mbsub</span>$<span class="no">endage</span>)</pre></body></html></div>
<p>A note on singletons: for all per-individual measures unconnected individuals within the pedigree are automatically excluded from the calculations as no kinship-based statistics can be estimated for them (they do, by definition, not share kinship with any other individual in the pedigree, thus their kinship coefficient with any other individual in the pedigree will be <code>0</code>). Note also that the removal of e.g. not phenotyped individuals prior to the calculation can also <em>generate</em> singletons, that additionally become removed. This removal results in an estimate with the value <code>NA</code> for all singletons as well as not phenotyped individuals.</p>
<p>Next, we calculate the mean FIR within each family and plot this information.</p>
<div class="sourceCode" id="cb109"><html><body><pre class="r"><span class="co">## Split the FIR by family and average the values within each.</span>
<span class="no">frFam</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="no">fr</span>, <span class="kw">f</span> <span class="kw">=</span> <span class="no">fad</span>$<span class="no">family</span>)
<span class="no">frFamAvg</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">frFam</span>, <span class="no">mean</span>, <span class="kw">na.rm</span> <span class="kw">=</span> <span class="fl">TRUE</span>)

<span class="co">## Sort and plot the averages.</span>
<span class="no">frFamAvg</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">frFamAvg</span>), <span class="kw">decreasing</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">frFamAvg</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"h"</span>, <span class="kw">xaxt</span> <span class="kw">=</span> <span class="st">"n"</span>, <span class="kw">xlab</span> <span class="kw">=</span> <span class="st">""</span>, <span class="kw">ylab</span> <span class="kw">=</span> <span class="st">"mean FIR"</span>,
     <span class="kw">main</span> <span class="kw">=</span> <span class="st">"Per family averaged familial incidence rate"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/axis.html">axis</a></span>(<span class="kw">side</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">las</span> <span class="kw">=</span> <span class="fl">2</span>, <span class="kw">at</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">frFamAvg</span>), <span class="kw">label</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">frFamAvg</span>))</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/fir-2,%20mbreast-mean-fr-per-family-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Not unexpectedly, individuals in some families have on average a higher familial incidence rate, and thus a higher risk of cancer than others.</p>
<p>In the next example, we calculate the familial incidence rate assessing in addition the significance of each estimate using Monte Carlo simulations. This extension to the original approach from Kerber <span class="citation">[6]</span> does also allow stratified sampling.</p>
<div class="sourceCode" id="cb110"><html><body><pre class="r"><span class="co">## Estimate the risk for each individual using the familial incidence</span>
<span class="co">## rate method. We use the endage provided in the Minnesota Breast Cancer</span>
<span class="co">## Record as a measure for time at risk.</span>
<span class="no">frTest</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData-analysis.html">familialIncidenceRateTest</a></span>(<span class="no">fad</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="no">tcancer</span>,
                    <span class="kw">traitName</span> <span class="kw">=</span> <span class="st">"cancer"</span>,
                    <span class="kw">timeAtRisk</span> <span class="kw">=</span> <span class="no">mbsub</span>$<span class="no">endage</span>,
                    <span class="kw">nsim</span> <span class="kw">=</span> <span class="no">nsim</span>)</pre></body></html></div>
<p>The familial incidence rate can be extracted easily from the result object using the <code>familialIncidenceRate</code> method or using <code>$fir</code>. Also, the empirical p-value from the simulation analysis and the time at risk can be accessed using the <code><a href="../reference/FAData.html">$</a></code> operator (i.e. using <code>$pvalue</code>, <code>$tar</code> or <code>$timeAtRisk</code>, respectively).</p>
<div class="sourceCode" id="cb111"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAData-analysis.html">familialIncidenceRate</a></span>(<span class="no">frTest</span>))</pre></body></html></div>
<pre><code>##           1           2           3           4           5           6 
##          NA 0.002278208 0.002365165 0.000670492 0.002709228 0.002098398</code></pre>
<div class="sourceCode" id="cb113"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">frTest</span>$<span class="no">fir</span>)</pre></body></html></div>
<pre><code>##           1           2           3           4           5           6 
##          NA 0.002278208 0.002365165 0.000670492 0.002709228 0.002098398</code></pre>
<p>Finally, we inspect the results from the analysis.</p>
<div class="sourceCode" id="cb115"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAProbResults.html">result</a></span>(<span class="no">frTest</span>))</pre></body></html></div>
<pre><code>##      trait_name total_phenotyped total_affected total_tested   id family
## 3185     cancer             3508            248         1778 3185     77
## 7122     cancer             3508            248         1778 7122    173
## 7125     cancer             3508            248         1778 7125    173
## 7123     cancer             3508            248         1778 7123    173
## 7124     cancer             3508            248         1778 7124    173
## 7121     cancer             3508            248         1778 7121    173
##              fir pvalue padj
## 3185 0.015789474      0    0
## 7122 0.010449918      0    0
## 7125 0.008874950      0    0
## 7123 0.008848773      0    0
## 7124 0.008587047      0    0
## 7121 0.008249860      0    0</code></pre>
<p>We can also identify the families containing individuals with a significant FIR.</p>
<div class="sourceCode" id="cb117"><html><body><pre class="r"><span class="no">frRes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAProbResults.html">result</a></span>(<span class="no">frTest</span>)
<span class="no">frSig</span> <span class="kw">&lt;-</span> <span class="no">frRes</span>[<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">frRes</span>$<span class="no">padj</span> <span class="kw">&lt;</span> <span class="fl">0.05</span>), ]

<span class="co">## Split by family.</span>
<span class="no">frFam</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="no">frSig</span>, <span class="no">frSig</span>$<span class="no">family</span>)
<span class="no">frRes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">family</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">frFam</span>),
            <span class="kw">no_sign_fir</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">frFam</span>, <span class="no">nrow</span>)))
<span class="co">## Determine the number of phenotyped and affected individuals per family.</span>
<span class="no">noPheNAff</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">frFam</span>), <span class="kw">function</span>(<span class="no">z</span>){
    <span class="no">fam</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData.html">family</a></span>(<span class="no">frTest</span>, <span class="kw">family</span> <span class="kw">=</span> <span class="no">z</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">no_pheno</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">fam</span>$<span class="no">affected</span>)),
         <span class="kw">no_aff</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">fam</span>$<span class="no">affected</span> <span class="kw">==</span> <span class="fl">1</span>))
         ))
})
<span class="no">frRes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="no">frRes</span>, <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="no">noPheNAff</span>))

<span class="co">## Display the number of phenotyped and affected individuals as well as</span>
<span class="co">## the number of individuals within the families with a significant FIR.</span>
<span class="no">frRes</span>[<span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="no">frRes</span>[, <span class="st">"no_sign_fir"</span>], <span class="kw">decreasing</span> <span class="kw">=</span> <span class="fl">TRUE</span>), ]</pre></body></html></div>
<pre><code>##     family no_sign_fir no_pheno no_aff
## 432    432          12      123     15
## 173    173           8       35     10
## 77      77           1       68      5</code></pre>
<p>We have an enrichment of affected cases in families 173, 13 and 432.</p>
</div>
<div id="kinship-sum-test" class="section level2">
<h2 class="hasAnchor">
<a href="#kinship-sum-test" class="anchor"></a><em>Kinship sum test</em> <a id="orgb323644"></a>
</h2>
<p>Next, we use the <em>kinship sum test</em> that evaluates familial aggregation based on the sum of kinship values between affected cases. The test identifies affected individuals exhibiting a higher relationship to other affected individuals than would be expected by chance. By specifying the <code>strata</code> we perform sex-stratified random sampling, i.e. ensure that the proportion of male and female individuals in each randomly sampled group matches the corresponding proportions in the <em>real</em>, observed, affected.</p>
<div class="sourceCode" id="cb119"><html><body><pre class="r"><span class="co">## Perform the kinship sum test.</span>
<span class="no">kinSum</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData-analysis.html">kinshipSumTest</a></span>(<span class="no">fad</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="no">tcancer</span>, <span class="kw">traitName</span> <span class="kw">=</span> <span class="st">"cancer"</span>,
             <span class="kw">nsim</span> <span class="kw">=</span> <span class="no">nsim</span>, <span class="kw">strata</span> <span class="kw">=</span> <span class="no">fad</span>$<span class="no">sex</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAProbResults.html">result</a></span>(<span class="no">kinSum</span>))</pre></body></html></div>
<pre><code>##       trait_name total_phenotyped total_affected affected_id family affected
## 17528     cancer             3508            248       17528    432      245
## 17517     cancer             3508            248       17517    432      245
## 17529     cancer             3508            248       17529    432      245
## 17547     cancer             3508            248       17547    432      245
## 17548     cancer             3508            248       17548    432      245
## 17549     cancer             3508            248       17549    432      245
##       kinship_sum         freq       pvalue        padj
## 17528        2.00 0.0004081633 5.714286e-05 0.004714286
## 17517        1.75 0.0040816327 1.346939e-04 0.004714286
## 17529        1.75 0.0040816327 1.346939e-04 0.004714286
## 17547        1.75 0.0040816327 1.346939e-04 0.004714286
## 17548        1.75 0.0040816327 1.346939e-04 0.004714286
## 17549        1.75 0.0040816327 1.346939e-04 0.004714286</code></pre>
<p>Next, we identify those individuals that have a significant kinship sum accepting a 10% false discovery rate (FDR).</p>
<div class="sourceCode" id="cb121"><html><body><pre class="r"><span class="co">## Extract the IDs of the individuals with significant kinship. By default,</span>
<span class="co">## the raw p-values are adjusted for multiple hypothesis testing using the</span>
<span class="co">## method from Benjamini and Hochberg.</span>
<span class="no">kinSumRes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAProbResults.html">result</a></span>(<span class="no">kinSum</span>)
<span class="no">kinSumIds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">kinSumRes</span>[<span class="no">kinSumRes</span>$<span class="no">padj</span> <span class="kw">&lt;</span> <span class="fl">0.1</span>, <span class="st">"affected_id"</span>])

<span class="co">## From which families are these?</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">kinSumRes</span>[<span class="no">kinSumIds</span>, <span class="st">"family"</span>])</pre></body></html></div>
<pre><code>## 
## 173 432 
##   6  12</code></pre>
<p>Thus, most of the identified significant individuals are from two families. Next, we compare the FIR scores of affected or unaffected (but phenotyped) individuals in this family to the FIR scores of affected or unaffected individuals of all other families.</p>
<div class="sourceCode" id="cb123"><html><body><pre class="r"><span class="co">## Get the familial ratio of the significant in this family, of all in</span>
<span class="co">## this family, and of all others.</span>
<span class="no">famId</span> <span class="kw">&lt;-</span> <span class="no">kinSumRes</span>[<span class="fl">1</span>, <span class="st">"family"</span>]

<span class="co">## Extract the family.</span>
<span class="no">fam</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData.html">family</a></span>(<span class="no">kinSum</span>, <span class="kw">family</span> <span class="kw">=</span> <span class="no">famId</span>)

<span class="co">## Stratify individuals in affected/unaffected.</span>
<span class="no">strat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"All, unaff."</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">kinSum</span>$<span class="no">id</span>))
<span class="no">strat</span>[<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">kinSum</span>$<span class="no">affected</span> <span class="kw">&gt;</span> <span class="fl">0</span>)] <span class="kw">&lt;-</span> <span class="st">"All, aff."</span>
<span class="no">strat</span>[<span class="no">kinSum</span>$<span class="no">id</span> <span class="kw">%in%</span> <span class="no">fam</span>$<span class="no">id</span>] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"Fam "</span>, <span class="no">famId</span>, <span class="st">", unaff."</span>)
<span class="no">strat</span>[<span class="no">kinSum</span>$<span class="no">id</span> <span class="kw">%in%</span> <span class="no">fam</span>$<span class="no">id</span>[<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">fam</span>$<span class="no">affected</span> <span class="kw">&gt;</span> <span class="fl">0</span>)]] <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"Fam "</span>,<span class="no">famId</span>,<span class="st">", aff."</span>)

<span class="no">famData</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">fr</span> <span class="kw">=</span> <span class="no">fr</span>, <span class="kw">group</span> <span class="kw">=</span> <span class="no">strat</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span>(<span class="no">fr</span>~<span class="no">group</span>, <span class="kw">data</span> <span class="kw">=</span> <span class="no">famData</span>, <span class="kw">na.rm</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">ylab</span> <span class="kw">=</span> <span class="st">"FIR"</span>,
    <span class="kw">col</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#FBB4AE"</span>, <span class="st">"#B3CDE3"</span>), <span class="fl">2</span>))</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/kinsum-3,%20mbreast-family-432-FIR-compared-to-others-1.png" width="700" style="display: block; margin: auto;"></p>
<p>As expected, the familial incidence rate (i.e., in the present data set, the risk of individuals to get cancer, given their kinship to other cancer cases) for individuals (whether affected or yet unaffected) in this family is higher than in the data set analyzed here.</p>
<p>Next, we plot the pedigree of this family.</p>
<div class="sourceCode" id="cb124"><html><body><pre class="r"><span class="co">## Plot the pedigree for the family of the selected individual removing</span>
<span class="co">## all individuals that were not phenotypes.</span>
<span class="fu"><a href="../reference/FAData.html">plotPed</a></span>(<span class="no">kinSum</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="no">kinSumIds</span>[<span class="fl">1</span>], <span class="kw">cex</span> <span class="kw">=</span> <span class="fl">0.3</span>, <span class="kw">only.phenotyped</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/kinsum-4,%20mbreast-family-432-affected-1.png" width="700" style="display: block; margin: auto;"></p>
<p>And finally, also plot the kinship sum for the individuals with the largest kinship sum in relation to the <em>expected</em> kinship sums from the Monte Carlo simulations.</p>
<div class="sourceCode" id="cb125"><html><body><pre class="r"><span class="fu"><a href="../reference/FAGenIndexResults.html">plotRes</a></span>(<span class="no">kinSum</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="no">kinSumIds</span>[<span class="fl">1</span>])</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/kinsum-5,%20mbreast-family-432-affecte-res-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div id="kinship-group-test" class="section level2">
<h2 class="hasAnchor">
<a href="#kinship-group-test" class="anchor"></a><em>Kinship group test</em> <a id="orgaac6b02"></a>
</h2>
<p>Here we apply the <em>kinship group test</em> to the data set. This test first defines for each affected individual a group of individuals considering only individuals that are as closely related as the most distant affected individual. For each of these kinship groups two tests are then performed, one by comparing the mean kinship among affected in the group with the mean kinship from Monte Carlo simulations (ratio test) and one evaluating the largest observed kinship value between affected individuals with those of random samples from the simulation (kinship group test).</p>
<p>In the example below we specify again the <code>strata</code> argument and thus perform sex-stratified random sampling.</p>
<div class="sourceCode" id="cb126"><html><body><pre class="r"><span class="co">## Calculate the kinship test.</span>
<span class="no">kinGroup</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData-analysis.html">kinshipGroupTest</a></span>(<span class="no">fad</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="no">tcancer</span>,
                 <span class="kw">traitName</span> <span class="kw">=</span> <span class="st">"cancer"</span>,
                 <span class="kw">nsim</span> <span class="kw">=</span> <span class="no">nsim</span>, <span class="kw">strata</span> <span class="kw">=</span> <span class="no">fad</span>$<span class="no">sex</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/FAProbResults.html">result</a></span>(<span class="no">kinGroup</span>))</pre></body></html></div>
<pre><code>##       trait_name total_phenotyped total_affected phenotyped affected group_id
## 410       cancer             3508            248       1147      174      410
## 2984      cancer             3508            248       1147      174     2984
## 17609     cancer             3508            248       1147      174    17609
## 7117      cancer             3508            248       1147      174     7117
## 17517     cancer             3508            248       1147      174    17517
## 17547     cancer             3508            248       1147      174    17547
##       family group_phenotyped group_affected ratio_pvalue ratio_padj
## 410       13                8              5            0          0
## 2984      72                1              2            0          0
## 17609    432                6              5            0          0
## 7117     173               19              8            0          0
## 17517    432               53             13            0          0
## 17547    432               55             14            0          0
##       mean_kinship kinship_pvalue kinship_padj
## 410      0.2500000              0            0
## 2984     0.2500000              0            0
## 17609    0.2500000              0            0
## 7117     0.1607143              0            0
## 17517    0.1458333              0            0
## 17547    0.1346154              0            0</code></pre>
<p>The kinship group test finds a significant aggregation of cases in families 13, 72, 173 and 432. In fact, as we see further below, the test identified a subgroup in the latter which shows with an exceptional high proportion of cases.</p>
<p>Below, we summarize the results further by listing the total number of families in the pedigree and the number of families in which kinship groups with significant kinship p-value and significant ratio p-value (both at a 5% FDR).</p>
<div class="sourceCode" id="cb128"><html><body><pre class="r"><span class="no">kinGroupRes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAProbResults.html">result</a></span>(<span class="no">kinGroup</span>)
<span class="co">## Create a data.frame with the summarized results.</span>
<span class="no">resTab</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">total_families</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">kinGroup</span>$<span class="no">family</span>)),
             <span class="kw">ratio_sign</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(
             <span class="no">kinGroupRes</span>[<span class="no">kinGroupRes</span>$<span class="no">ratio_padj</span> <span class="kw">&lt;</span> <span class="fl">0.05</span>, <span class="st">"family"</span>]
             )),
             <span class="kw">kinship_sign</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(
             <span class="no">kinGroupRes</span>[<span class="no">kinGroupRes</span>$<span class="no">kinship_padj</span> <span class="kw">&lt;</span> <span class="fl">0.05</span>, <span class="st">"family"</span>]
             ))
             )
<span class="no">resTab</span></pre></body></html></div>
<pre><code>##   total_families ratio_sign kinship_sign
## 1             69          6            9</code></pre>
<p>The most significant kinship group identified by the kinship group test is shown in the figure below. The mother (individual <code>17609</code>) of the nuclear family representing this group and all her daughters have cancer (see figure below). This mother is however not directly related to the affected founder of this family, individual <code>17517</code>, but did marry her son (id <code>17530</code>; see figure above for the full pedigree of this family <code>432</code>).</p>
<p>We are also submitting the familial incidence rate values calculated above with argument <code>label1</code> which are then displayed below the ID of each individual in the plot.</p>
<div class="sourceCode" id="cb130"><html><body><pre class="r"><span class="fu"><a href="../reference/FAData.html">plotPed</a></span>(<span class="no">kinGroup</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="no">kinGroupRes</span>[<span class="no">kinGroupRes</span>$<span class="no">family</span> <span class="kw">==</span> <span class="st">"432"</span>,
                   <span class="st">"group_id"</span>][<span class="fl">1</span>],
    <span class="kw">prune</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">label1</span> <span class="kw">=</span> <span class="no">fr</span>)</pre></body></html></div>
<p><img src="FamAgg_files/figure-html/kingroup-3,%20mbreast-family-432-affecte-res-kinship-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div id="binomial-test" class="section level2">
<h2 class="hasAnchor">
<a href="#binomial-test" class="anchor"></a>Binomial test</h2>
<p>The binomial test evaluates whether the number of affected in a family (or the whole pedigree) is significantly higher than what would be expected by chance (given a probability of being affected in a trait). In contrast to most other methods this test does not take the degree of kinship between individuals into account and is hence independent of the family structure in the pedigree. We can perform this type of test using the <code>binomialTest</code> function on any <code>FAData</code> object or any object extending it. Below we use the binomial test to evaluate a significant enrichment of affected individuals in any family in the pedigree.</p>
<div class="sourceCode" id="cb131"><html><body><pre class="r"><span class="no">binRes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData-analysis.html">binomialTest</a></span>(<span class="no">fad</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="no">tcancer</span>, <span class="kw">traitName</span> <span class="kw">=</span> <span class="st">"Cancer"</span>)

<span class="no">binResTab</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAProbResults.html">result</a></span>(<span class="no">binRes</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">binResTab</span>)</pre></body></html></div>
<pre><code>##     trait_name total_phenotyped total_affected family phenotyped affected
## 173     Cancer             3508            248    173         35       10
## 19      Cancer             3508            248     19         24        5
## 432     Cancer             3508            248    432        123       15
## 94      Cancer             3508            248     94         36        6
## 8       Cancer             3508            248      8         37        6
## 14      Cancer             3508            248     14         32        5
##           pvalue       prob        padj
## 173 0.0001101636 0.07069555 0.007601286
## 19  0.0241694261 0.07069555 0.603487360
## 432 0.0273283250 0.07069555 0.603487360
## 94  0.0388827384 0.07069555 0.603487360
## 8   0.0437309681 0.07069555 0.603487360
## 14  0.0720884731 0.07069555 0.766407020</code></pre>
<p>The probability used on the binomial test is shown in column <code>"prob"</code> and is in essence the ratio between the affected and phenotyped in the pedigree (i.e. 154/2202). This might be an overestimation, especially if the provided pedigree is not representative of the population. A population-based probability can however be provided with argument <code>prob</code>. Below we test specifically whether we have families in which the number of individuals with breast cancer is significantly higher than expected. To this end we set the trait status of all male individuals to <code>NA</code> and repeat the test providing the probability of developing breast cancer during in women, which, according to the U.S. Breast Cancer Statistics (from breastcancer.org) is 1 out of 8 in their life time.</p>
<div class="sourceCode" id="cb133"><html><body><pre class="r"><span class="co">## Set the trait status to NA for all male individuals.</span>
<span class="no">tcancer</span>[<span class="no">fad</span>$<span class="no">sex</span> <span class="kw">==</span> <span class="st">"M"</span> <span class="kw">|</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">fad</span>$<span class="no">sex</span>)] <span class="kw">&lt;-</span> <span class="fl">NA</span>

<span class="co">## Perform the test providing also the population probability</span>
<span class="no">binRes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAData-analysis.html">binomialTest</a></span>(<span class="no">fad</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="no">tcancer</span>, <span class="kw">prob</span> <span class="kw">=</span> <span class="fl">1</span>/<span class="fl">8</span>)

<span class="no">binResTab</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/FAProbResults.html">result</a></span>(<span class="no">binRes</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">binResTab</span>)</pre></body></html></div>
<pre><code>##     trait_name total_phenotyped total_affected family phenotyped affected
## 14          NA             1990            206     14         15        5
## 19          NA             1990            206     19         12        4
## 13          NA             1990            206     13         18        5
## 94          NA             1990            206     94         18        5
## 8           NA             1990            206      8         19        5
## 173         NA             1990            206    173         19        5
##         pvalue  prob      padj
## 14  0.03107294 0.125 0.9090792
## 19  0.05281048 0.125 0.9090792
## 13  0.06464965 0.125 0.9090792
## 94  0.06464965 0.125 0.9090792
## 8   0.07905037 0.125 0.9090792
## 173 0.07905037 0.125 0.9090792</code></pre>
<p>Below we plot the pedigree for the family with the strongest enrichment with affected individuals.</p>
<div class="sourceCode" id="cb135"><html><body><pre class="r"><span class="fu"><a href="../reference/FAData.html">plotPed</a></span>(<span class="no">binRes</span>, <span class="kw">family</span> <span class="kw">=</span> <span class="fl">173</span>)</pre></body></html></div>
<pre><code>## Warning in kinship2::pedigree(id = individual, dadid = father, momid = mother, :
## More than 25% of the gender values are 'unknown'</code></pre>
<p><img src="FamAgg_files/figure-html/bintest-3-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code>## Did not plot the following people: 7135 7141 7143 7144 7145 7146 7148 7149 26800 26811 26812 26813</code></pre>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Sinnwell:2014kd">
<p>1. Sinnwell JP, Therneau TM, Schaid DJ: <strong>The kinship2 R package for pedigree data.</strong> <em>Human heredity</em> 2014, <strong>78</strong>:91–93.</p>
</div>
<div id="ref-Rainer:2016hk">
<p>2. Rainer J, Taliun D, D’Elia Y, Pattaro C, Domingues FS, Weichenberger CX: <strong>FamAgg: an R Package to Evaluate Familial Aggregation of Traits in Large Pedigrees.</strong> <em>Bioinformatics</em> 2016:btw019.</p>
</div>
<div id="ref-Thiele:2005">
<p>3. Thiele H, Nuernberg P: <strong>HaploPainter: A tool for drawing pedigrees with complex haplotypes.</strong> <em>Bioinformatics</em> 2005, <strong>21</strong>:1730–1732.</p>
</div>
<div id="ref-Purcell:2007dg">
<p>4. Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MAR, Bender D, Maller J, Sklar P, Bakker PIW de, Daly MJ, Sham PC: <strong>PLINK: a tool set for whole-genome association and population-based linkage analyses.</strong> <em>American journal of human genetics</em> 2007, <strong>81</strong>:559–575.</p>
</div>
<div id="ref-Hill:1980tz">
<p>5. Hill JR: <strong>A kinship survey of cancer in the Utah Mormon population</strong>. In <em>Cancer incidence in defined populations. Banbury report 4.</em> Edited by Cairns MJ, Lyon JL, Skolnic M. Cold Spring Harbor, NY; 1980:299–318.</p>
</div>
<div id="ref-Kerber:1995cx">
<p>6. Kerber RA: <strong>Method for calculating risk associated with family history of a disease.</strong> <em>Genetic epidemiology</em> 1995, <strong>12</strong>:291–301.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
